# Hadamard and Phase Gate Bloch-Sphere Visualization

## 1. Problem Statement
Create a 1-qubit quantum circuit initialized in $|0\rangle$, apply a Hadamard gate followed by a Phase ($S$) gate, obtain the resulting statevector, and visualize the final quantum state on the Bloch sphere.

---

## 2. Theoretical Background & Concepts

### Initial State and the Hadamard Gate
In quantum computing, single qubits are initialized by default in the computational basis state $|0\rangle$. Applying a **Hadamard gate** ($H$) creates an equal superposition of basis states:

$$H|0\rangle = \frac{|0\rangle + |1\rangle}{\sqrt{2}} = |+\rangle$$

In vector representation:
$$|+\rangle = \frac{1}{\sqrt{2}}\begin{pmatrix} 1 \\ 1 \end{pmatrix}$$

In this state, both computational basis states have equal amplitudes of $\frac{1}{\sqrt{2}}$, with zero relative phase ($\phi = 0$).

### The Phase (S) Gate
The **Phase gate** ($S$, or $\sqrt{Z}$) is a single-qubit unitary gate that introduces a phase shift of $\pi/2$ (a factor of $e^{i\pi/2} = i$) specifically to the $|1\rangle$ component, leaving $|0\rangle$ unaffected:

$$S = \begin{pmatrix} 1 & 0 \\ 0 & e^{i\pi/2} \end{pmatrix} = \begin{pmatrix} 1 & 0 \\ 0 & i \end{pmatrix}$$

Applying the $S$ gate to the equal superposition state $|+\rangle$ yields:

$$S|+\rangle = S\left(\frac{|0\rangle + |1\rangle}{\sqrt{2}}\right) = \frac{|0\rangle + i|1\rangle}{\sqrt{2}} = |+i\rangle$$

The resulting statevector is complex-valued:

$$|\psi\rangle = \begin{pmatrix} \frac{1}{\sqrt{2}} \\ \frac{i}{\sqrt{2}} \end{pmatrix} \approx \begin{pmatrix} 0.7071 \\ 0.7071i \end{pmatrix}$$

### Geometric Representation on the Bloch Sphere
The **Bloch sphere** provides a geometric representation of any single-qubit pure state on the surface of a unit 3D sphere:

$$|\psi\rangle = \cos\left(\frac{\theta}{2}\right)|0\rangle + e^{i\phi}\sin\left(\frac{\theta}{2}\right)|1\rangle$$

where:
- $\theta \in [0, \pi]$ is the polar angle, defining the population balance between $|0\rangle$ (North pole, $\theta = 0$) and $|1\rangle$ (South pole, $\theta = \pi$).
- $\phi \in [0, 2\pi)$ is the azimuthal angle, representing the relative phase in the equatorial plane.

### Effect of the S Gate on the Bloch Sphere
1. **Initial State $|0\rangle$**: Positioned at the North pole along the $+Z$ axis ($\theta = 0$).
2. **After Hadamard Gate ($|+\rangle$)**: The state vector rotates to the equator along the positive $X$-axis ($\theta = \pi/2, \phi = 0$).
3. **After Phase Gate ($|+i\rangle$)**: The $S$ gate applies a rotation of $\pi/2$ ($90^\circ$) around the $Z$-axis, rotating the state vector from the $+X$ axis to the positive $Y$-axis along the equator ($\theta = \pi/2, \phi = \pi/2$).
4. **Measurement Probabilities vs. Relative Phase**: Because the state remains on the equator ($\theta = \pi/2$), computational basis measurement probabilities are unchanged:
   $$P(0) = \left|\frac{1}{\sqrt{2}}\right|^2 = 0.5, \quad P(1) = \left|\frac{i}{\sqrt{2}}\right|^2 = 0.5$$
   The $S$ gate modifies solely the relative phase $\phi$, which is geometrically visualized as an equatorial rotation on the Bloch sphere.

---

## 3. Circuit Flow
The sequence of operations is:

$$|0\rangle \xrightarrow{\quad \text{Hadamard (H)} \quad} |+\rangle = \frac{|0\rangle + |1\rangle}{\sqrt{2}} \xrightarrow{\quad \text{Phase (S)} \quad} |\psi\rangle = \frac{|0\rangle + i|1\rangle}{\sqrt{2}}$$

---

## 4. Tools & Libraries Used
- **Qiskit**: IBM's open-source quantum computing SDK for defining and manipulating quantum circuits (`QuantumCircuit`).
- **`qiskit.quantum_info.Statevector`**: Analytical tool for inspecting exact statevectors without measurement collapse.
- **Qiskit Visualization (`plot_bloch_multivector`)**: Specialized plotting utility for rendering quantum statevectors on the Bloch sphere.
- **Matplotlib**: Python plotting framework used to display graphic figures.

### Step 1: Import Required Libraries
Import `QuantumCircuit`, `Statevector`, `plot_bloch_multivector` from Qiskit, and `matplotlib.pyplot` for visualization.

In [1]:
from qiskit import QuantumCircuit
from qiskit.quantum_info import Statevector
from qiskit.visualization import plot_bloch_multivector
import matplotlib.pyplot as plt


### Step 2: Create the Single-Qubit Circuit
Initialize a 1-qubit quantum circuit in the default computational ground state $|0\rangle$.

In [2]:
# Create circuit
qc = QuantumCircuit(1)


### Step 3: Apply the Hadamard Gate
Apply the Hadamard gate (`qc.h(0)`) to create the equal superposition state $|+\rangle$.

In [3]:
# Apply Hadamard
qc.h(0)


### Step 4: Apply the Phase (S) Gate
Apply the Phase gate (`qc.s(0)`) to rotate the relative phase of the state by $\pi/2$.

In [4]:
# Apply Phase (S) gate
qc.s(0)


### Step 5: Obtain the Statevector
Extract the exact quantum statevector from the circuit instructions using `Statevector.from_instruction(qc)` and display both the circuit diagram and statevector.

In [5]:
# Statevector
state = Statevector.from_instruction(qc)

print("Quantum Circuit:")
print(qc)

print("\nStatevector:")
print(state)


Quantum Circuit:
   ┌───┐┌───┐
q: ┤ H ├┤ S ├
   └───┘└───┘

Statevector:
Statevector([0.70710678+0.j        , 0.        +0.70710678j],
            dims=(2,))


### Step 6: Visualize the State on the Bloch Sphere
Render the final quantum statevector on the Bloch sphere using `plot_bloch_multivector` to observe its equatorial position along the $+Y$ axis.

In [6]:
# Bloch Sphere
plot_bloch_multivector(state)
plt.show()


## Conclusion

- **Superposition and Phase Evolution**: The experiment demonstrates the complementary roles of the Hadamard and Phase gates in single-qubit state engineering:
  - The **Hadamard gate** transforms the ground state $|0\rangle$ into the equal superposition state $|+\rangle$, placing the Bloch vector on the equator along the $+X$ axis.
  - The **Phase (S) gate** advances the relative phase of the $|1\rangle$ component by $\pi/2$, rotating the Bloch vector by $90^\circ$ around the $Z$-axis into the $+Y$ direction ($|+i\rangle$).
- **Phase Modification vs. Measurement Probabilities**: When acting on $|+\rangle$, the $S$ gate does **not** change computational basis measurement probabilities—the probabilities remain balanced at $P(0) = |1/\sqrt{2}|^2 = 0.5$ and $P(1) = |i/\sqrt{2}|^2 = 0.5$. Instead, it specifically modifies the quantum relative phase $\phi$.
- **Bloch-Sphere Geometric Representation**: The Bloch-sphere visualization and complex statevector `[0.70710678+0.j, 0.+0.70710678j]` provide an intuitive geometric representation of how phase rotations manipulate quantum states orthogonally to computational basis populations.